dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala

```diff
@@ -0,0 +1,176 @@
+            val (partitionKeyFilters, dataFilter) =
+              getPartitionKeyFiltersAndDataFilters(
+                fsRelation.sparkSession,
+                relation,
+                partitionSchema,
+                filters,
+                relation.output)
+            val prunedPartitionSize = fileIndex.listFiles(
+              partitionKeyFilters.toSeq,
+              dataFilter)
+              .size
+            if (prunedPartitionSize > maxScanPartitions) {
+              throw maxPartitionExceedError(
+                prunedPartitionSize,
+                maxScanPartitions,
+                relation.catalogTable,
+                fileIndex.rootPaths,
+                fsRelation.partitionSchema)
+            }
+            Nil
+          case ScanOperation(
+                _,
+                filters,
+                logicalRelation @ LogicalRelation(
+                  fsRelation @ HadoopFsRelation(
+                    catalogFileIndex: CatalogFileIndex,
+                    partitionSchema,
+                    _,
+                    _,
+                    _,
+                    _),
+                  _,
+                  _,
+                  _)) if fsRelation.partitionSchemaOption.isDefined =>
+            val (partitionKeyFilters, _) =
+              getPartitionKeyFiltersAndDataFilters(
+                fsRelation.sparkSession,
+                logicalRelation,
+                partitionSchema,
+                filters,
+                logicalRelation.output)
+
+            val prunedPartitionSize =
+              catalogFileIndex.filterPartitions(
+                partitionKeyFilters.toSeq)
+                .partitionSpec()
+                .partitions
+                .size
+            if (prunedPartitionSize > maxScanPartitions) {
+              throw maxPartitionExceedError(
+                prunedPartitionSize,
+                maxScanPartitions,
+                logicalRelation.catalogTable,
+                catalogFileIndex.rootPaths,
+                fsRelation.partitionSchema)
+            }
+            Nil
+          case _ => Nil
+        }
+      case _ => Nil
+    }
+  }
+
+  def maxPartitionExceedError(
+      prunedPartitionSize: Int,
+      maxPartitionSize: Int,
+      tableMeta: Option[CatalogTable],
+      fileIndex: Seq[Path],
+      partitionSchema: StructType): Throwable = {
+    new MaxPartitionExceedException(
+      s"""
+         |SQL job scan data source partition: $prunedPartitionSize
+         |exceed restrict of data source scan maxPartition $maxPartitionSize
+         |You should optimize your SQL logical according partition structure
+         |or shorten query scope such as p_date, detail as below:
+         |Table: ${tableMeta.map(_.qualifiedName).getOrElse("")}
+         |Owner: ${tableMeta.map(_.owner).getOrElse("")}
+         |RootPaths: ${fileIndex.mkString(",")}
+         |Partition Structure: ${partitionSchema.map(_.name).mkString(" -> ")}
```

Comment:
can you provide an example how this msg looks like ?

Reply:
- textfile hive table w/ partition key filters
```
SQL job scan hive partition: 9
exceed restrict of hive scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
Partition Structure: p
```

- textfile hive table w/o partition key filters
```
Your SQL job scan a whole huge table without any partition filter,
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
Partition Structure: p
```

- parquet hive table
```
SQL job scan data source partition: 10
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-c59b6bb7-2c66-4dd5-9774-4bf60850e995/org.apache.spark.sql.WatchDogSuite/warehouse/test
Partition Structure: p
```

- data source

```
SQL job scan data source partition: 6
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: 
Owner: 
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-7a07fd59-c94f-46cf-8bf6-bbeb2a494697
Partition Structure: p
```

Reply:
what happens if there are more one partitions ?

Reply:
```
SQL job scan data source partition: 9
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=0,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=2,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=3,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=4,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=5... 4 more paths
Partition Structure: p
```

Reply:
```scala
    withTempDir { dir =>
      withTempView("test") {
        spark.range(10)
          .selectExpr("id", "explode(sequence(0, 10)) as v")
          .selectExpr("id", "id as p", "v as p2")
          .write
          .partitionBy("p", "p2")
          .mode("overwrite")
          .save(dir.getCanonicalPath)
        spark.read.load(dir.getCanonicalPath).createOrReplaceTempView("test")
        withSQLConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS.key -> "5") {
          sql("SELECT * FROM test where p=1").queryExecution.sparkPlan
        }
      }
    }
```

```
SQL job scan data source partition: 11
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: 
Owner: 
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-cbda132a-e5b0-4f66-849a-f566b730f7ec
Partition Structure: p -> p2
```

Reply:
`Partition Structure: p -> p2`
should it be ?
`Partition Structure: p, p2`

Reply:
Yeah, updated.